### PR TITLE
guestfs: flag to set uid and gid

### PR DESCRIPTION
**Pull request by alicefr**
Signed-off-by: Alice Frosi <afrosi@redhat.com>

**What this PR does / why we need it**:
Add flag to specify uid for guestfs pod.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Flag for setting the guestfs uid and gid
```

### CODE REVIEW

**Review by xpivarc — COMMENTED** [review]
/hold
So here are the possibilities:
1. User creating a guestfs Pod has no `MustRunAsRange` for the user. In this case, openshift will not allocate the uid for you and you need to set it.
2. User creating a guestfs Pod has `MustRunAsRange` and in this case, the uid should be allocated.

Can we confirm you see the 1. case?

**Comment by alicefr** [comment]
@xpivarc is there a way from the user to check which scc they are allow to use?

**Comment by alicefr** [comment]
I changed this to allow the user to specify the uid. This PR preserves the old default behavior. If no root or uid flag is specified the uid is let empty and it will rely on the cluster default

**Comment by alicefr** [comment]
But I'd really like to find a nice way to automatically set the guestfs pod uid.

**Review by xpivarc — COMMENTED** [review]
Looks good to me. Let's add explicit validation that uid must not be set when root flag is set.

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by xpivarc** [comment]
A test covering this would also be good to have.

**Comment by alicefr** [comment]
Yes, let me add a unit test to cover this

**Comment by alicefr** [comment]
@xpivarc or do we want to have also a functional test with running under a different uid?

**Comment by xpivarc** [comment]
> @xpivarc is there a way from the user to check which scc they are allow to use?

You could potentially dry-run the pod and see if it was rejected by scc admission. Personally, I think the solution we discussed is better than doing this.

File: tests/storage/guestfs.go [hunk 1]
```diff
@@ -102,6 +102,18 @@ var _ = SIGDescribe("[rfe_id:6364][[Serial]Guestfs", func() {
 
 	}
 
+	verifyCanRunOnFSPVC := func(podName string) {
+		stdout, stderr, err := execCommandLibguestfsPod(podName, []string{"qemu-img", "create", "/disk/disk.img", "500M"})
+		Expect(stderr).To(Equal(""))
+		Expect(stdout).To(ContainSubstring("Formatting"))
+		Expect(err).ToNot(HaveOccurred())
+		stdout, stderr, err = execCommandLibguestfsPod(podName, []string{"guestfish", "-a", "/disk/disk.img", "run"})
+		Expect(stderr).To(Equal(""))
```
Comment (xpivarc) [inline tests/storage/guestfs.go:111]: While at it you could use `BeEmpty`. +1 for this small refactor

File: pkg/virtctl/guestfs/guestfs.go [hunk 2]
```diff
@@ -368,12 +374,55 @@ func (client *K8sClient) getPodsForPVC(pvcName, ns string) ([]corev1.Pod, error)
...
+		if err != nil {
+			return nil, err
+		}
+		return &n, nil
+	default:
+		return nil, nil
+	}
+}
+
+func setGIDroupLibguestfs() (*int64, error) {
```
Comment (vasiliy-ul) [inline pkg/virtctl/guestfs/guestfs.go:415]: The function name: shouldn't it be `setGIDLibguestfs`?

**Comment by vasiliy-ul** [comment]
Also, the tests probably need to be adapted: https://github.com/kubevirt/kubevirt/blob/7c395bb0654aec24af968ef3fd4aceff1d88a878/tests/storage/guestfs.go#L80

**Comment by rmohr** [comment]
/test pull-kubevirt-e2e-kind-1.22-sriov

unblocking this test. There seems to have been a hickup, where no job ever got craeted. This blockes also the retets bot because PRs are stuck in pending.

**Comment by alicefr** [comment]
@xpivarc could you please have a look to this when you have a bit of time?

File: pkg/virtctl/guestfs/guestfs.go [hunk 3]
```diff
@@ -368,12 +374,55 @@ func (client *K8sClient) getPodsForPVC(pvcName, ns string) ([]corev1.Pod, error)
...
+		}
+		return &n, nil
+	default:
+		return nil, nil
+	}
+}
+
+func setGIDLibguestfs() (*int64, error) {
+	// The GID can only be specified together with the uid. See comment at: https://github.com/kubernetes/cri-api/blob/2b5244cefaeace624cb160d6b3d85dd3fd14baea/pkg/apis/runtime/v1/api.proto#L307-L309
+	if gid != "" && uid == "" {
```
Comment (xpivarc) [inline pkg/virtctl/guestfs/guestfs.go:417]: For convenience, we could match gid to uid.

Comment (alicefr) [reply pkg/virtctl/guestfs/guestfs.go:417]: don't we have situations where the uid might be different from 0 but the gid is still 0?

Comment (alicefr) [reply pkg/virtctl/guestfs/guestfs.go:417]: I'd like to keep them disjoint for situations where you have files on the PVC that requires a specific group and the user wants to control those values.

**Comment by brianmcarey** [comment]
/override pull-kubevirt-e2e-k8s-1.24-sig-storage-dv-gc

Lane has been removed as it is no longer required.

**Pull request merged**